### ISSUE TITLE
How to refine LLM result based on human feedback: `Reinforcement learning from human feedback(RLHF)`

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#22]
![image](https://github.com/user-attachments/assets/9fa740df-a31c-4e50-96fe-4ae138248304)

---

### Context:

- we find out that in one-shot case, not-fine-tuned llm model(chatgpt 3o free) can not give a valid result for generating image processing code
- we would like to refine the llm result in a recursive way

----

### Totally different usecases:

- though our intended usecase is `input : text description for processing image + image` and `output: workflow composed of operators in json format `
- the expertise of the text description varies, could be more `technical` or more `subjective`

---

**- Usecase 1: technical user descriptions to technical operations**

![image](https://github.com/user-attachments/assets/2d2c91c0-1a52-4c78-8511-9bd0319b201c)

- this text description has valid technical information
- this usecase has a clear right/wrong answer (the workflow either works or doesn't)
-  supervised learning approach could work well 
  
---

**- Usecase 2: non-technical, subjective user descriptions to technical operations**

![image](https://github.com/user-attachments/assets/9673fed0-735c-48f9-a832-c1a6584096f8)

```
- multiple workflows could work:

- Workflow A: [sharpen → color boost → contrast]
- Workflow B: [denoise → HDR effect → saturation → vignette]
- Workflow C: [auto-levels → vibrance → selective contrast]

```
- this text description is less technical but more subjective
- multiple workflows are `correct` (valid JSON, executable), but some might be:

      - More computationally efficient
      - Produce more aesthetically pleasing results
      - Use fewer steps while achieving similar results
      - Better match current social media trends

**Comment by xp1632** [comment]
### Usecase 1: technical user descriptions to technical operations

![image](https://github.com/user-attachments/assets/2d2c91c0-1a52-4c78-8511-9bd0319b201c)

- this text description has valid technical information
- this usecase has a clear right/wrong answer (the workflow either works or doesn't)
- a supervised learning approach could work well 
----

### How to refine the result of Usecase 1: 

**1. Create a high-quality training dataset:**
        - Collecting pairs of natural language descriptions and their corresponding working JSON workflows
        - Having experts create "golden" examples of optimal workflows for different descriptions
        - Mining logs of successful user interactions with your platform


**2. Fine-tuning an existing LLM on a training dataset**

**3. Using few-shot prompting with golden examples**

**4. Adding a validation step to ensure generated JSON follows operator/platform's constraints**

**Comment by xp1632** [comment]
**- Usecase 2: non-technical, subjective user descriptions to technical operations**

![image](https://github.com/user-attachments/assets/9673fed0-735c-48f9-a832-c1a6584096f8)

```
- multiple workflows could work:

- Workflow A: [sharpen → color boost → contrast]
- Workflow B: [denoise → HDR effect → saturation → vignette]
- Workflow C: [auto-levels → vibrance → selective contrast]

```
- this text description is less technical but more subjective
- multiple workflows are `correct` (valid JSON, executable), but some might be:

      - More computationally efficient
      - Produce more aesthetically pleasing results
      - Use fewer steps while achieving similar results
      - Better match current social media trends

---

### How to refine the result of Usecase 2: 


- as this usecase involves user's subjective `preference`
- if we want the llm result to better align the `preference` based on user feedback 
- we need `Reinforcement learning from human feedback`

**Comment by xp1632** [comment]
### What's Reinforcement learning (RL)

- While `supervised learning` learn by **correct labeled training data**
- `Reinforcement learning` learns through **trial, error, rewards** 
- In RL,  an agent interacts with an environment and receives feedback in the form of **rewards** or **penalties** based on its actions.
---
![image](https://github.com/user-attachments/assets/0537790c-9edb-4528-8609-f84e90f8a2c6)
![image](https://github.com/user-attachments/assets/10d5a2c9-eada-4d01-8edb-044ba5204419)


- **Concrete example of tic tac toe** 
- **Objective** :  win the game 
- Agent first takes a random `action` which leads to a new state and collects the responding reward by this action
- after several iterative rounds of interaction with the environment, the agent learns the optimal policy for a given environment that maximizes rewards
- agent accumulates experience by gradually uncovering actions that yield the highest long-term rewards, ultimately leading to success in the game

**Comment by xp1632** [comment]
### How can Reinforcement Learning fine-tune LLM with human feedback?

![image](https://github.com/user-attachments/assets/6e38f5c1-c9fe-47bf-8e2b-a68706f89a04)

- **Objective:** Generate text that is aligned with human preference 
- **Agent:** Instruct LLM model
- **Environment:** LLM context window of the model, texts are entered by a prompt
- **State:** is the current text contained in the context window
- **Action:** is the act of generating text
- **Reward** is given when the generated text aligns with human preference

---

### What's the function of `reward model`?

- while use a human expert to give scalar feedback for every action can be time-consuming and expensive
- we can train an additional reward model as a scalable alternative instead
![image](https://github.com/user-attachments/assets/5174f0ee-d848-4e56-a98a-8c9b2233095e)

- `Reward model` classify the llm outputs and evaluate the degree of alignment with human preferences

- we start with smaller number of human examples to train reward model by traditional supervised learning
- after the reward model is trained , we use this secondary model to assess the output of llm and assign `reward` value, and this updates the weights of LLM to get a more human-aligned version

**Comment by xp1632** [comment]
### How to train a `reward model` used in RLHF?

- we put a pause here today since rlhf might be too complicated for our demand
- let's dive into how to deploy llm then quick review this part
